### PR TITLE
HDDS-9020. Datanodes fails to start up when secret key has not yet been initialized in SCM.

### PR DESCRIPTION
**Pull request by duongkame**
## What changes were proposed in this pull request?

When a datanode/OM starts up while SCM has not finished initializing secret keys yet, the startup fails because Datanode/OM needs to prefetch the current active secret keys.

```
2023-07-13 16:06:22,369 [main] ERROR org.apache.hadoop.ozone.HddsDatanodeService: Exception in HddsDatanodeService.
java.lang.RuntimeException: Can't start the HDDS datanode plugin
	at org.apache.hadoop.ozone.HddsDatanodeService.start(HddsDatanodeService.java:361)
	at org.apache.hadoop.ozone.HddsDatanodeService.start(HddsDatanodeService.java:235)
	at org.apache.hadoop.ozone.HddsDatanodeService.call(HddsDatanodeService.java:203)
	at org.apache.hadoop.ozone.HddsDatanodeService.call(HddsDatanodeService.java:93)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2346)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2311)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at org.apache.hadoop.hdds.cli.GenericCli.execute(GenericCli.java:100)
	at org.apache.hadoop.hdds.cli.GenericCli.run(GenericCli.java:91)
	at org.apache.hadoop.ozone.HddsDatanodeService.main(HddsDatanodeService.java:185)
Caused by: org.apache.hadoop.hdds.security.exception.SCMSecretKeyException: Secret key initialization is not finished yet.
	at org.apache.hadoop.hdds.protocolPB.SecretKeyProtocolClientSideTranslatorPB.handleError(SecretKeyProtocolClientSideTranslatorPB.java:101)
	at org.apache.hadoop.hdds.protocolPB.SecretKeyProtocolClientSideTranslatorPB.submitRequest(SecretKeyProtocolClientSideTranslatorPB.java:89)
	at org.apache.hadoop.hdds.protocolPB.SecretKeyProtocolClientSideTranslatorPB.getCurrentSecretKey(SecretKeyProtocolClientSideTranslatorPB.java:127)
	at org.apache.hadoop.hdds.security.symmetric.DefaultSecretKeySignerClient.start(DefaultSecretKeySignerClient.java:72)
	at org.apache.hadoop.hdds.security.symmetric.DefaultSecretKeyClient.start(DefaultSecretKeyClient.java:50)
	at org.apache.hadoop.ozone.HddsDatanodeService.start(HddsDatanodeService.java:312)
	... 13 more
2023-07-13 16:06:22,382 [shutdown-hook-0] INFO org.apache.hadoop.ozone.HddsDatanodeService: SHUTDOWN_MSG: 
/************************************************************
SHUTDOWN_MSG: Shutting down HddsDatanodeService at ....
************************************************************/  
```

The current active secret key is mandatory for Datanode/OM startup because they need it in their background processes, e.g. EC reconstruction. The solution is to apply retries to the active secret key prefetch if SCM has not initialized secret keys yet.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9020

Please replace this section with the link to the Apache JIRA)

## How was this patch tested?

I added a custom code to SCM to delay the secret key initialization, started a docker cluster and verified datanodes/OMs retry on SCM secret key exceptions until succeed.

```
2023-07-14 17:45:18,748 [main] INFO symmetric.DefaultSecretKeyVerifierClient: Initializing secret key cache with size 16, TTL PT168H
2023-07-14 17:45:18,875 [main] INFO utils.RetriableTask: Execution of task getCurrentSecretKey failed, will be retried in 4000 ms
2023-07-14 17:45:22,882 [main] INFO utils.RetriableTask: Execution of task getCurrentSecretKey failed, will be retried in 6000 ms
2023-07-14 17:45:28,895 [main] INFO utils.RetriableTask: Execution of task getCurrentSecretKey failed, will be retried in 19000 ms
2023-07-14 17:45:47,938 [main] INFO utils.RetriableTask: Execution of task getCurrentSecretKey failed, will be retried in 32000 ms
2023-07-14 17:46:19,972 [main] INFO utils.RetriableTask: Execution of task getCurrentSecretKey failed, will be retried in 91000 ms
2023-07-14 17:47:51,007 [main] INFO symmetric.DefaultSecretKeySignerClient: Initial secret key fetched from SCM: SecretKey(id = f937b502-faca-4505-a894-72c19dde972b, creation at: 2023-07-14T17:47:16.371Z, expire at: 2023-07-21T17:47:16.371Z).
```

### CODE REVIEW

**Review by errose28 — APPROVED** [review]
LGTM. IMO retry with exponential backoff up to a fixed time like this PR uses is better than the infinite retry being used in places like `HAUtils#getCAListWithRetry`. The later can cause the process to hang indefinitely with little clear feedback to the user, which we have seen in clusters before.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/DefaultSecretKeySignerClient.java [hunk 1]
```diff
@@ -68,13 +74,46 @@ public void refetchSecretKey() {
...
     scheduleSecretKeyPoller(conf, initialKey.getCreationTime());
   }
 
+  private ManagedSecretKey loadInitialSecretKey() throws IOException {
+    // Load initial active secret key from SCM, retries with exponential
+    // backoff when SCM has not initialized secret keys yet.
+
+    // Exponential backoff policy, 10 retries/1s will give maximum wait time
+    // around 10 min (2^9 = 512s).
+    int maxRetries = 10;
```
Comment (kerneltime) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/DefaultSecretKeySignerClient.java:90]: Process state != initialized. Datanode can be up indefinitely in a secure cluster for SCM to be initialized to return the secret keys, but the process need not stop.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/DefaultSecretKeySignerClient.java [hunk 2]
```diff
@@ -68,13 +74,46 @@ public void refetchSecretKey() {
 
   @Override
   public void start(ConfigurationSource conf) throws IOException {
-    final ManagedSecretKey initialKey =
-        secretKeyProtocol.getCurrentSecretKey();
+    final ManagedSecretKey initialKey = loadInitialSecretKey();
```
Comment (kerneltime) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/DefaultSecretKeySignerClient.java:77]: What should happen if block tokens are not enabled? Maybe Datanode should not wait for SCM to start up.

Comment (duongkame) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/DefaultSecretKeySignerClient.java:77]: This path is reachable only when block tokens are enabled.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/DefaultSecretKeySignerClient.java (hunk 1 above)
Comment (duongkame) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/DefaultSecretKeySignerClient.java]: Increased the max retries to 100, with exp backoff wait time repeats after 10, the total maximum wait time is around 3h.

**Pull request merged**